src/plugins/data/common/search/search_source/legacy/fetch_soon.ts

```diff
@@ -83,20 +83,27 @@ async function delayedFetch(
   if (ms === 0) {
     return callClient([request], [options], fetchHandlers)[0];
   }
+  const { dataSourceId, ...restRequest } = request;
+  const key = dataSourceId ?? '';
 
-  const i = requestsToFetch.length;
-  requestsToFetch = [...requestsToFetch, request];
-  requestOptions = [...requestOptions, options];
+  const i = requestsToFetch[key]?.length ?? 0;
+  requestsToFetch[key] = [...(requestsToFetch[key] ?? []), restRequest];
+  requestOptions[key] = [...(requestOptions[key] ?? []), options];
```

Comment:
Nit: better to add some comments here to elaborate that we are grouping request by data source.